protos/orchestrator_service.proto

```diff
@@ -270,6 +270,7 @@ message CreateInstanceRequest {
     OrchestrationIdReusePolicy orchestrationIdReusePolicy = 6;
     google.protobuf.StringValue executionId = 7;
     map<string, string> tags = 8;
+    TraceContext parentTraceContext = 9;
```

Comment:
I'm a little bit surprised that we need this. I've been running some tests in DTS and I'm seeing that the distributed trace ID is flowing as expected, even without this change. Let me investigate that further and get back to you.

Reply:
I think the issue is out of proc worker enqueueing orchestrations. I haven't tested it myself, but there is probably one of two issues happening:

1. Correlation is flowing at all. As it is a process hop, the `Activity.Current` the user has in their worker isn't available to the host process, so it doesn't capture the correct parent trace context when enqueueing.
2. OR, it is capturing it, but there is are extra gRPC client & server spans in between representing the call from worker to host to schedule this orchestration. This is _sometimes_ okay, sometimes not. When the user owns the server handing the enqueueing it is okay and subjective on if you like those extra spans or not. When the user _doesn't_ own the server (DTS), this may be an issue as the user may not be collecting the gRPC server span, leading to a broken/incomplete dependency graph.

Reply:
I did some testing, and it seems that the .NET client/server seem to know how to automatically flow the distributed trace context across gRPC calls. in my test, I started a distributed trace on the client, prior to calling `ScheduleNewOrchestrationInstanceAsync`. The trace ID was:

> 00-0c5e28bfd6b73a79675b7a48e63a888a-5c1cc52c35ddabb7-01

I then checked the value of `Activity.Current` on the gRPC server implementation, and it showed the following trace ID:

> 00-0c5e28bfd6b73a79675b7a48e63a888a-3e8ac421f295a014-01

Since the trace ID is the same and only the span ID is different, I concluded that HttpClient and ASP.NET seem to be handling this trace propagation for us automatically. I therefore worry if we might end up generating inconsistent or disconnected spans if there are two separate mechanisms for generating a child span and flowing it to the server.

@bachuv I wonder if we should try a similar test with the .NET Isolated worker?

> When the user owns the server handing the enqueueing it is okay and subjective on if you like those extra spans or not. When the user doesn't own the server (DTS), this may be an issue as the user may not be collecting the gRPC server span, leading to a broken/incomplete dependency graph.

Not sure I totally followed this reasoning. In our case, Microsoft owns both the client (Durable Task SDK) and the server (DTS or the Durable Functions WebJobs extension). What would the ideal outcome be?

Reply:
> Not sure I totally followed this reasoning. In our case, Microsoft owns both the client (Durable Task SDK) and the server (DTS or the Durable Functions WebJobs extension). What would the ideal outcome be?

The issue isn't Microsoft owning both. It is the _customer_ being able to capture the telemetry from both. We don't want service owned spans (DTS in this case) to be inserted in the middle of a customer owned flow.

Take service bus distributed tracing for example. Using the SB SDK, you get spans that look like this:

<some parent span> -> SB producer span (from the SDK) -> SB consumer span (from the SDK again) -> any further spans the customer handler code emits.

Even though this message is sent to the SB service between the produce/consumer span, the spans from the SB service are not involved.

Additionally, even if we did want DTS's gRPC service span involved here, the technical feasibility of implementing that is essentially a non-starter.

Reply:
The actual problem is still fuzzy to me since I don't understand how customers would be able to collect spans that are generated by the server. But I don't see this as a blocking issue. I can see how this would be valuable for languages/stacks (other than .NET) where trace context doesn't flow automatically.

Reply:
> I don't understand how customers would be able to collect spans that are generated by the server

That is exactly the issue this is addressing. Customers typically cannot collect those spans. This PR explicitly flows the captured span the customer does emit, so the backend can persist that as part of the orchestration metadata. This then lets the SDK use that trace context as the parent when the orchestration is ran on a worker.

Reply:
> @bachuv I wonder if we should try a similar test with the .NET Isolated worker?

I know this comment is resolved, but I wanted to add my observations after testing. I started an `Activity` before calling `ScheduleNewOrchestrationInstanceAsync` and the `Activity.Current` value is accurate until we get to [`StartInstance()` in LocalGrpcListener.cs](https://github.com/Azure/azure-functions-durable-extension/blob/53abf0f5d893340e2ebeb24e352377c5d0c15393/src/WebJobs.Extensions.DurableTask/LocalGrpcListener.cs#L153). It's value is null at that point. With adding the trace context to `P.CreateInstanceRequest`, we can recreate the `Activity` again in `StartInstance()` using the `P.CreateInstanceRequest.TraceContext` information. Let me know if you would like me to test another scenario.